packages/typescript-estree/src/parseSettings/ExpiringCache.ts

```diff
@@ -0,0 +1,111 @@
+export const DEFAULT_TSCONFIG_CACHE_DURATION_SECONDS = 30;
+
+/**
+ * A map with key-level expiration.
+ */
+export class ExpiringCache<TKey, TValue> {
+  readonly #cacheDurationSeconds: number | 'Infinity';
+  /**
+   * The mapping of path-like string to the resolved TSConfig(s)
+   */
+  protected map = new Map<
+    TKey,
+    Readonly<{
+      value: TValue;
+      lastSeen: [number, number];
+    }>
+  >();
+
+  constructor(cacheDurationSeconds: number | 'Infinity') {
+    this.#cacheDurationSeconds = cacheDurationSeconds;
+  }
+
+  set(key: TKey, value: TValue): void {
+    this.map.set(key, {
+      value,
+      lastSeen:
+        this.#cacheDurationSeconds === 'Infinity'
+          ? // no need to waste time calculating the hrtime in infinity mode as there's no expiry
+            [0, 0]
+          : process.hrtime(),
+    });
+  }
+
+  get(key: TKey): TValue | undefined {
+    const entry = this.map.get(key);
+    if (entry?.value != null) {
+      const ageSeconds = process.hrtime(entry.lastSeen)[0];
+      if (
+        this.#cacheDurationSeconds === 'Infinity' ||
+        ageSeconds < this.#cacheDurationSeconds
+      ) {
+        // cache hit woo!
+        return entry.value;
+      } else {
+        // key has expired - clean it up to free up memory
+        this.cleanupKey(key);
+      }
+    }
+    // no hit :'(
+    return undefined;
+  }
+
+  protected cleanupKey(key: TKey): void {
+    this.map.delete(key);
+  }
+
+  get size(): number {
+    return this.map.size;
+  }
+
+  clear(): void {
+    this.map.clear();
+  }
+}
+
+/**
+ * A map with key-level expiration where it is known that there will be multiple
+ * keys with the same value.
+ */
+export class ExpiringCacheWithReverseMap<TKey, TValue> extends ExpiringCache<
```

Comment:
This isn't used anywhere. Dead code, to be removed?

Reply:
I added this ahead of your PR (#6084) just because.
Can remove it for you to copy pasta over there if you'd like!

Reply:
Heh, serves you right for trying to be helpful to me 😄. Good to know - I can copy & paste it from the early commit here. But yeah let's remove for now.